### PR TITLE
7 as a user i want to identify all token holders that held my nft for a given time period

### LINKED ISSUE #7: As a User, I want to identify all token holders that held my NFT for a given time period
Acceptance criteria:
- This should be a feature of the token-holders-list
- There is a new input that takes a duration. The duration will be used to check if a given account has held X amount of NFTs for at least the duration time.
- The duration is optional, the user can still provide just tokenId and amount.
- There is a radio button to select if the user wants to provide a date or provide a duration in days/weeks/months. Days/weeks/months should be provided as an integer
  - Day = 24 hours
  - Week = 24 hours * 7
  - Month = 24 hours * 30
- The app returns all accountIds that are holding **at least X** NFTs from tokenId for at least a given duration at the time of execution (no need to calculate the whole history of ownership for each token, just the current state)
- Unit and end2end tests written

Check just the current token holders. No need to create a whole history of transfers and ownership

It should use this endpoint to check for transfers
https://mainnet-public.mirrornode.hedera.com/api/v1/tokens/0.0.3838071/nfts/1/transactions

### PR DESCRIPTION
**Pull request by StanislawMazowieckiAriane**

### CODE REVIEW

File: src/utils/fetchNftsWithDuration.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import { NftsResponseType } from '@/types/nfts-response';
+import { getHolderAndDuration } from 'nft-utilities/src/getHolderAndDuration/index';
+import { changeDurationToDate } from '@/utils/changeDurationToDate';
+
+export const fetchNftsWithDuration = async (nftBalances: BalancesWithNFT[]): Promise<BalancesWithNFT[]> => {
+  for (const nftBalance of nftBalances) {
+    // Skip if the NFT doesn't have a duration. Duration is not required
+    if (!nftBalance.duration) continue;
+
+    let countSuccessfullyNfts = 0;
```
Comment (BartoszSolkaArianelabs) [inline src/utils/fetchNftsWithDuration.ts]: let's name it `nftsMeetingDurationCondition`

File: src/utils/changeDurationToDate.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  if (typeof duration === 'string') {
+    const now = new Date();
+
+    switch (durationType) {
+      case 'days':
+        return new Date(now.setDate(now.getDate() - Number(duration)));
+      case 'weeks':
+        return new Date(now.setDate(now.getDate() - Number(duration) * 7));
+      case 'months':
+        return new Date(now.setMonth(now.getMonth() - Number(duration)));
```
Comment (BartoszSolkaArianelabs) [inline src/utils/changeDurationToDate.ts]: a month should be always treated as 30 days in our case

File: src/utils/isValidTokenId.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+export const isValidTokenId = (tokenId: string): boolean => {
+  const regex = /^0\.0\.\d*$/;
+  return regex.test(tokenId);
```
Comment (BartoszSolkaArianelabs) [inline src/utils/isValidTokenId.ts]: https://docs.hedera.com/hedera/sdks-and-apis/hedera-api/basic-types/tokenid

using `0.0.` prefix will work for now, but a more general approach would be better

**Review by BartoszSolkaArianelabs — CHANGES_REQUESTED** [review]

File: src/utils/fetchNftsWithDuration.ts (hunk 1 above)
Comment (StanislawMazowieckiAriane) [reply src/utils/fetchNftsWithDuration.ts]: Done

File: src/utils/changeDurationToDate.ts (hunk 2 above)
Comment (StanislawMazowieckiAriane) [reply src/utils/changeDurationToDate.ts]: Done

File: src/utils/isValidTokenId.ts (hunk 3 above)
Comment (StanislawMazowieckiAriane) [reply src/utils/isValidTokenId.ts]: Done

**Comment by BartoszSolkaArianelabs** [comment]
1. optimize checking `Break the loop if the count of successfully fetched NFTs is greater than or equal to the minAmount`.
2. Switch for date/duration is changing background color
3. Text for switch for date/duration is not aligned with state of the switch 
image

**Review by michalrozekariane — APPROVED** [review]

**Review by BartoszSolkaArianelabs — APPROVED** [review]

**Pull request merged**